ee/billing/billing_manager.py

```diff
@@ -52,6 +53,61 @@ def handle_billing_service_error(res: requests.Response, valid_codes=(200, 404,
             raise Exception(f"Billing service returned bad status code: {res.status_code}", f"body:", res.text)
 
 
+def compute_usage_per_tier(current_usage: int, projected_usage: int, tiers):
+    remaining_usage = current_usage
+    remaining_projected_usage = projected_usage or 0
+    previous_tier: Optional[dict[str, Any]] = None
+    tier_max_usage: Union[int, float] = 0
+
+    result: list[Tier] = []
+    for tier in tiers:
+        if previous_tier and previous_tier.get("up_to"):
+            previous_tier_up_to = previous_tier["up_to"]
+        else:
+            previous_tier_up_to = 0
+
+        if tier.get("up_to"):
+            tier_max_usage = tier["up_to"] - previous_tier_up_to
+        else:
+            tier_max_usage = float("inf")
+
+        flat_amount_usd = Decimal(tier.get("flat_amount_usd") or 0)
+        unit_amount_usd = Decimal(tier.get("unit_amount_usd") or 0)
+        usage_this_tier = int(min(remaining_usage, tier_max_usage))
+        remaining_usage -= usage_this_tier
+        current_amount_usd = Decimal(unit_amount_usd * usage_this_tier + flat_amount_usd).quantize(Decimal("0.01"))
+        previous_tier = tier
+        if projected_usage:
+            projected_usage_this_tier = int(min(remaining_projected_usage, tier_max_usage))
+            remaining_projected_usage -= projected_usage_this_tier
+            projected_amount_usd = Decimal(unit_amount_usd * projected_usage_this_tier + flat_amount_usd).quantize(
+                Decimal("0.01")
+            )
+        else:
+            projected_usage_this_tier = None
+            projected_amount_usd = None
+
+        result.append(
+            Tier(
+                flat_amount_usd=str(flat_amount_usd),
```

Comment:
String coercion is annoying but it prevents upstream problems in the UI as all of these values are typed as strings.